src/ynEIGEN/EigenStrategyManager.sol

```diff
@@ -234,27 +317,49 @@ contract EigenStrategyManager is
 
         node.depositAssetsToEigenlayer(depositAssets, depositAmounts, strategiesForNode);
 
+        for (uint256 i = 0; i < assetsLength; i++) {
+            updateTokenStakingNodesBalances(assets[i], IStrategy(address(0)));
+        }
+
         emit DepositedToEigenlayer(depositAssets, depositAmounts, strategiesForNode);
     }
 
-    function toEigenLayerDeposit(
-        IERC20 asset,
-        uint256 amount
-    ) internal returns (IERC20 depositAsset, uint256 depositAmount) {
-        if (address(asset) == address(wstETH)) {
-            // Adjust for wstETH
-            depositAsset = stETH;
-            depositAmount = wstETH.unwrap(amount); 
-        } else if (address(asset) == address(woETH)) {
-            // Adjust for woeth
-            depositAsset = oETH; 
-            // calling redeem with receiver and owner as address(this)
-            depositAmount = woETH.redeem(amount, address(this), address(this)); 
-        } else {
-            // No adjustment needed
-            depositAsset = asset;
-            depositAmount = amount;
-        }   
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  WITHDRAWALS  -------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    function processPrincipalWithdrawals(
+        WithdrawalAction[] calldata _actions
+    ) public onlyRole(WITHDRAWAL_MANAGER_ROLE)  {
+        uint256 _len = _actions.length;
+        for (uint256 i = 0; i < _len; ++i) {
+            _processPrincipalWithdrawalForNode(_actions[i]);
+        }
+    }
+
+    function _processPrincipalWithdrawalForNode(WithdrawalAction calldata _action) internal {
+
+        uint256 _totalAmount = _action.amountToReinvest + _action.amountToQueue;
+
+        ITokenStakingNode _node = tokenStakingNodesManager.getNodeById(_action.nodeId);
+        _node.deallocateTokens(IERC20(_action.asset), _totalAmount);
+        IERC20(_action.asset).safeTransferFrom(address(_node), address(this), _totalAmount);
+
+        if (_action.amountToReinvest > 0) {
+            IynEigen _ynEigen = ynEigen;
+            IERC20(_action.asset).forceApprove(address(_ynEigen), _action.amountToReinvest);
+            _ynEigen.processWithdrawn(_action.amountToReinvest, _action.asset);
+        }
+
+        if (_action.amountToQueue > 0) {
+            IRedemptionAssetsVaultExt _redemptionAssetsVault = redemptionAssetsVault;
+            IERC20(_action.asset).forceApprove(address(_redemptionAssetsVault), _action.amountToQueue);
+            _redemptionAssetsVault.deposit(_action.amountToQueue, _action.asset);
+        }
+
+        updateTokenStakingNodesBalances(IERC20(_action.asset), IStrategy(address(0)));
```

Comment:
we can move this after the loop of processPrincipalWithdrawals

and have it only be computed once